mdassemble.8
.\" -*- nroff -*-
.TH MDASSEMBLE 8 "" v3.4
.SH NAME
mdassemble \- assemble MD devices
.I aka
Linux Software RAID

.SH SYNOPSIS

.BI mdassemble

.SH DESCRIPTION
.B mdassemble
is a tiny program that can be used to assemble MD devices inside an
initial ramdisk (initrd) or initramfs; it is meant to replace the in-kernel
automatic RAID detection and activation.
It can be built statically and linked against lightweight libc alternatives, like
.B dietlibc,
.B klibc
or
.B uClibc.

.SH USAGE
Invoking
.B mdassemble
has the same effect as invoking
.B mdadm \-\-assemble \-\-scan.
.PP
Invoking
.B mdassemble
a second time will make all defined arrays readwrite, this is useful if
using the
.B start_ro
module parameter.

.SH OPTIONS

There are no options to
.B mdassemble.

.SH FILES

.SS /etc/mdadm.conf

The config file lists which devices may be scanned to see if
they contain MD super block, and gives identifying information
(e.g. UUID) about known MD arrays.  See
.BR mdadm.conf (5)
for more details.

.B mdassemble
supports all configuration parameters defined in
.B mdadm.conf
with the exception of
.B auto=
which is supported only if mdadm was built with the
.B \-DMDASSEMBLE_AUTO
define.

.SH SEE ALSO
.PP
.BR mdadm (8),
.BR mdadm.conf (5),
.BR md (4),
.BR diet (1).


ChangeLog
Please see git logs for detailed change log.
This file just contains highlight.

Changes Prior to release 3.1.4
  Two fixes related to configs that aren't using udev:
   -    Don't remove md devices which 'standard' names on --stop
   -    Allow dev_open to work on read-only /dev
  And fixed regressions:
   -    Allow --incremental to add spares to an array
   -    Accept --no-degraded as a deprecated option rather than
            throwing an error
   -    Return correct success status when --incrmental assembling 
        a container which does not yet have enough devices.
   -    Don't link mdadm with pthreads, only mdmon needs it.
   -    Fix compiler warning due to bad use of snprintf

Changes Prior to release 3.1.3
   -    mapfile now lives in a fixed location which default to
        /dev/.mdadm/map but can be changed at compile time.  This
	location is choses and most distros provide it during early
	boot and preserve it through.  As long a /dev exists and is
	writable, /dev/.mdadm will be created.
	Other files file communication with mdmon live here too.
	This fixes a bug reported by Debian and Gentoo users where
	udev would spin in early-boot.
   -    IMSM and DDF metadata will not be recognised on partitions
        as they should only be used on whole-disks.
   -    Various overflows causes by 2G drives have been addressed.
   -    A subarray of an IMSM contain can now be killed with
        --kill-subarray.  Also subarrays can be renamed with
	--update-subarray
   -    -If (or --incremental --fail) can be used  from udev to
        fail and remove from all arrays a device which has been
	unplugged from the system.  i.e. hot-unplug-support.
   -    "mdadm /dev/mdX --re-add missing" will look for any device
        that looks like it should be a member of /dev/mdX but isn't
	and will automatically --re-add it
   -    Now compile with -Wextra to get extra warnings.
   -    Lots of minor bug fixes, documentation improvements, etcc

Changes Prior to release 3.1.2
   -    The default metadata has change again (sorry about that).
        It is now v1.2 and will hopefully stay that way.  It turned
        out there with boot-block issues with v1.1 which make it 
        unsuitable for a default, though in many cases it is still
        suitable to use.
   -    Stopping a container is not permitted when members are still
        active
   -    Add 'homehost' to the valid words for the "AUTO" config file
        line.  When followed by "-all", this causes mdadm to
        auto-assemble any array belonging to this host, but not
        auto-assemble anything else.
   -    Fix some bugs with "--grow --chunksize=" for changing chunksize.
   -    VAR_RUN can be easily changed at compile time just like ALT_RUN.
        This gives distros more flexability in how to manage the
        pid and sock files that mdmon needs.
   -    Various mdmon fixes
   -    Alway make bitmap 4K-aligned if at all possible.
   -    If mdadm.conf lists arrays which have inter-dependencies,
        the previously had to be listed in the "right" order.  Now
        any order should work.
   -    Fix --force assembly of v1.x arrays which are in the process
        of recovering.
   -    Add section on 'scrubbing' to 'md' man page.
   -    Various command-line-option parsing improvements.
   -    ... and lots of other bug fixes.
	
Changes Prior to release 3.1.1
   -    Multiple fixes for new --grow levels including fixes for
	serious data corruption problems.
   -    Change default metadata to v1.1
   -    Change default chunk size to 512K
   -    Change default bitmap chunk size to 64Meg
   -    When --re-add is used, don't fall back to
	--add if --re-add fails as this can destroy data.

Changes Prior to release 3.1
   -    Support --grow to change the layout of RAID4/5/6
   -    Support --grow to change the chunksize of raid 4/5/6
   -    Support --grow to change level from RAID1 -> RAID5 -> RAID6 and
        back.
   -    Support --grow to reduce the number of devices in RAID4/5/6.
   -    Support restart of these grow options which assembling an array 
	which is partially grown.
   -    Assorted tests of this code, and of different RAID6 layouts.
	
Changes Prior to release 3.0.3
   -    Improvements for creating arrays giving just a name, like 'foo',
	rather than the full '/dev/md/foo'.
   -    Improvements for assembling member arrays of containers.
   -    Improvements to test suite
   -    Add option to change increment for RebuildNN messages reported
	by "mdadm --monitor"
   -    Improvements to mdmon 'hand-over' from initrd to final root.
   -    Handle merging of devices that have left an IMSM array and are
	being re-incorporated.
   -    Add missing space in "--detail --brief" output.
	
Changes Prior to release 3.0.2
   -    Fix crash when hosthost is not set, as often happens in
	early boot.

Changes Prior to release 3.0.1
   -    Fix various segfaults
   -    Fixed for --examine with containers
   -    Lots of other little fixes.
	
Changes Prior to release 3.0
   -    Support for externally managed metadata, specifically DDF and IMSM.
   -    Depend on udev to create entries in /dev, rather than creating them
	ourselves.
   -    remove --auto-update-home-hosts
   -    new config file line "auto"
   -    new "<ignore>" and "any" options for "homehost"
   -    numerous bug fixes and minor enhancements.


mdassemble.c
/*
 * mdassemble - assemble Linux "md" devices aka RAID arrays.
 *
 * Copyright (C) 2001-2009 Neil Brown <neilb@suse.de>
 * Copyright (C) 2003 Luca Berra <bluca@vodka.it>
 *
 *
 *    This program is free software; you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program; if not, write to the Free Software
 *    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *    Author: Neil Brown
 *    Email: <neilb@suse.de>
 */

#include "mdadm.h"
#include "md_p.h"

char const Name[] = "mdassemble";

#ifndef MDASSEMBLE_AUTO
/* from mdopen.c */
int open_mddev(char *dev, int report_errors/*unused*/)
{
	int mdfd = open(dev, O_RDWR);
	if (mdfd < 0)
		pr_err("error opening %s: %s\n",
			dev, strerror(errno));
	else if (md_get_version(mdfd) <= 0) {
		pr_err("%s does not appear to be an md device\n",
			dev);
		close(mdfd);
		mdfd = -1;
	}
	return mdfd;
}
int create_mddev(char *dev, char *name, int autof/*unused*/, int trustworthy,
		 char *chosen)
{
	return open_mddev(dev, 0);
}
#endif

int rv;
int mdfd = -1;

int main(int argc, char *argv[])
{
	struct mddev_ident *array_list =  conf_get_ident(NULL);
	struct context c = { .freeze_reshape = 1 };
	if (!array_list) {
		pr_err("No arrays found in config file\n");
		rv = 1;
	} else
		for (; array_list; array_list = array_list->next) {
			mdu_array_info_t array;
			if (strcasecmp(array_list->devname, "<ignore>") == 0)
				continue;
			mdfd = open_mddev(array_list->devname, 0);
			if (mdfd >= 0 && ioctl(mdfd, GET_ARRAY_INFO, &array) == 0) {
				rv |= Manage_ro(array_list->devname, mdfd, -1); /* make it readwrite */
				continue;
			}
			if (mdfd >= 0)
				close(mdfd);
			rv |= Assemble(array_list->st, array_list->devname,
				       array_list, NULL, &c);
		}
	return rv;
}
